polyfill/index.d.ts

```diff
@@ -0,0 +1,385 @@
+     *
+     * In case of ambiguous or non-existent times, this option controls what
+     * absolute time to return:
+     * - `'earlier'`: The earlier time of two possible times
+     * - `'later'`: The later of two possible times
+     * - `'reject'`: Throw a RangeError instead.
+     *
+     * The default is `'earlier'`.
+     *
+     * Compatibility Note: the legacy `Date` object (and libraries like
+     * moment.js and Luxon) gives the same result as `earlier` when turning the
+     * clock back, and `later` when setting the clock forward.
+     *
+     * */
+    disambiguation: 'earlier' | 'later' | 'reject';
+    // TODO: should there be a `compatible` (to legacy `Date`, moment, etc.)
+    //       disambiguation option? See #614.
+  };
+
+  /**
+   * Options for arithmetic operations like `plus()` and `minus()`
+   * */
+  export type ArithmeticOptions = {
+    /**
+     * Controls handling of out-of-range arithmetic results.
+     *
+     * If a result is out of range, then `'constrain'` will clamp the result to
+     * the allowed range, while `'reject'` will throw a RangeError.
+     *
+     * The default is `'constrain'`.
+     */
+    disambiguation: 'constrain' | 'reject';
+    // TODO: may change name from `disambiguation` to `overflow`. See #607.
+  };
+
+  /**
+   * Options to control `Duration.prototype.minus()` behavior
+   * */
+  export type DurationMinusOptions = {
+    /**
+     * Controls how to deal with subtractions that result in negative overflows
+     *
+     * In `'balanceConstrain'` mode, negative fields are balanced with the next
+     * highest field so that none of the fields are negative in the result. If
+     * this is not possible, a `RangeError` is thrown.
+     *
+     * In `'balance'` mode, all fields are balanced with the next highest field,
+     * no matter if they are negative or not.
+     *
+     * TODO: add an example of each mode.
+     *
+     * The default is `'balanceConstrain'`.
+     *
+     */
+    disambiguation: 'balanceConstrain' | 'balance';
+  };
+
+  export interface DifferenceOptions<T extends string> {
+    /**
+     * The largest unit to allow in the resulting `Temporal.Duration` object.
+     *
+     * Valid values may include `'years'`, `'months'`, `'days'`, `'hours'`,
+     * `'minutes'`, and `'seconds'`, although some types may throw an exception
+     * if a value is used that would produce an invalid result. For example,
+     * `hours` is not accepted by `Date.prototype.difference()`.
+     *
+     * The default depends on the type being used.
+     *
+     * TODO: document which types use which defaults.
+     */
+    largestUnit: T;
+  }
+
+  export type DurationLike = {
+    years?: number;
+    months?: number;
+    days?: number;
+    hours?: number;
+    minutes?: number;
+    seconds?: number;
+    milliseconds?: number;
+    microseconds?: number;
+    nanoseconds?: number;
+  };
+
+  export class Duration extends Required<DurationLike> {
```

Comment:
The advantage of using extends is that it ties the DurationLike and Duration types together, so that if a change is made to one that doesn't match changes to the other (e.g. renaming a field) then you'll get a compiler error to catch the problem.

BTW, how to describe classes in TS is something I'm still not 100% sure about. I was going to research it today. One option is doing what's in the current commit: a `class` keyword with all type info inside. The other option is to create 2-3 types (one for the class constructor+statics, one for instance methods and fields, and an optional third `declare var` for global objects like `Array`), which essentially matches the old-school JS way to create classes. The second way of describing types is what TS's own standard types for ES built-in objects use.  So I'm trying to figure out if they only do it that way for backwards compat wtih ancient TS versions, or if there's some advantage of doing it that way. 

Depending on what I learn (see 
https://stackoverflow.com/questions/62006825/pros-cons-of-declaring-class-types-in-d-ts-using-class-foo-vs-separate-foo) it might make sense to change to the 2-to-3 type way before finalizing this PR.

Reply:
(Also a TypeScript noob here) Are you still going to get an error if the fields change, even if they are overridden readonly below?

Reply:
Oh, I didn't really answer your question.  Yes, it probably makes sense to add `Duration | DurationLike` to all places where we accept a `DurationLike` parameter.  In practice it won't change any type-checking behavior do make this change, but I agree with you that it would be clearer for folks who are reading the type info or viewing it in an IDE. 

I'll scrub all the use of `*Like` types and make this change.

Reply:
> @ptomato Are you still going to get an error if the fields change, even if they are overridden readonly below?

`readonly` refers to what the client code is allowed to do. It's what `get` properties look like in TS types. No problem if the underlying value changes--`readonly` just means that userland code isn't allowed to make those changes. If I try to write code that tries to change a readonly field, I'll get an error in the IDE and compiler. Like this:

![image](https://user-images.githubusercontent.com/277214/82836970-9e77b500-9e7c-11ea-8084-20f875ca0109.png)

Reply:
> It it necessary to use the `extends` clause? Can't we manually specify for each function that accepts a `DurationLike` that it may also accept a `Duration`?

One more thing: I just realized I was using `extends` but should have been using `implements`.  The former is used when you're inheriting from a class, while the latter is for when you're implementing an interface. I'll fix this along with the other issues noted above.

Reply:
Exactly. That's why I was confused, because you usually `extend` an existing class...

Reply:
> (Also a TypeScript noob here) Are you still going to get an error if the fields change, even if they are overridden readonly below?

@ptomato - to answer your question more directly, here's what you'll see if you try to rename a field in `Duration` without a  corresponding change in the `DurationLike` interface.

![image](https://user-images.githubusercontent.com/277214/82900562-9c136a80-9f11-11ea-86af-61a79a71744d.png)

Reply:
> Exactly. That's why I was confused, because you usually extend an existing class...

Apologies for misunderstanding your original comment.  I thought you were objecting to the dependency between the class vs. the interface, not to the incorrect `extends`. I almost never write classes in my own projects so I'm still adjusting to the quirks of TS when applied to classes. In class-less code, `extends` is the normal way to way to connect types together and `implements` is almost never used.  With classes it's the reverse.

Normally ESLint (which has great TS support now) would have caught this mistake but I discovered yesterday that my global ESLint config was ignoring .d.ts. This is a non-issue for my own projects where local ESLint config is part of the project, but caused false negatives in Temporal whose ESLint config ignores TS.

I'm looking into what it would require to add TS support to Temporal's ESLint config.  It's harder than I'd hoped because Temporal uses the babel-eslint parser which has more limited linting support than @typescript-eslint/parser. But if I can figure it out then I could add this change into this PR. I suspect the cleanest way to do it would be to create a separate TS-friendly eslint config and then to run ESLint twice in `npm run lint`, once with the regular config and again just for the .d.ts. Or to add a new script in package.json and run both of them in CI. Would something like this be welcome?

Reply:
The `babel-eslint` parser is used specifically because the default parser didn't support `import()`. You're welcome to switch it out for a different parser altogether as long as it lints the existing JS files cleanly. If not, then linting the types file with a separate config in `npm run lint` seems like a good way to go.

Reply:
I got @typescript-eslint/parser to work. It required upgrading eslint to 7.1 (because of https://stackoverflow.com/questions/54514989/) and turning off three lint rules that triggered false-positive warning that weren't real problems in the JS codebase.  But otherwise no changes were required. I'll add these changes to the PR.